jwt_auth/jwt.py

```
@@ -1,40 +1,70 @@
+import typing
+from django.conf import settings
 from typing import Optional, List
 
 import jwt
+from jwt_auth.jwks import Jwks
+
+# Load JSON Web Key Set globally, to cache the keys. Any service that
+# uses this library should restart to reset the key set.
+JWKS = Jwks()
 
 
 class JWT:
-    """Represents a JWT
-    """
+    """Represents a JWT."""
 
     def __init__(self, token: str, keys: List[str]):
         self._token = token
         self._keys = keys
+
         self._payload: Optional[dict] = None
 
     @property
     def payload(self) -> dict:
-        """The payload stored in the jwt
-        """
+        """The payload stored in the jwt."""
         if self._payload is None:
-            self._payload = self._decode()
+            # If RS256 header is detected, try to decode the token using
+            # JWKS if endpoint is configured.
+            keys = self._keys
+            unverified_header = jwt.get_unverified_header(self._token)
+            alg = unverified_header.get('alg')
+            if alg == 'RS256':
+                key = JWKS.get_jwk(unverified_header)
+                if key:
+                    # Use the fetched key because key is retrieved using
+                    # kid in the header, and therefore it should be valid.
+                    keys = [key]
```

if `keys` is an empty list at this point we should raise an informative error since we can't do it above anymore.